snapcraft/internal/lxd.py

```diff
@@ -316,18 +316,88 @@ def _ensure_container(self):
+        for interface in network:
+            for address in network[interface]['addresses']:
+                if address['family'] == 'inet':
+                    return address['address']
+        raise SnapcraftEnvironmentError('No IP found for {}'.format(
+            self._container_name))
+
     def _ensure_mount(self, destination, source):
         logger.info('Mounting {} into container'.format(source))
+        remote, container_name = self._container_name.split(':')
+        if remote != 'local':
+            return self._remote_mount(destination, source)
+
         devices = self._get_container_status()['devices']
         if destination not in devices:
             check_call([
                 'lxc', 'config', 'device', 'add', self._container_name,
                 destination, 'disk', 'source={}'.format(source),
                 'path={}'.format(destination)])
 
+    def _remote_mount(self, destination, source):
+        # Generate an SSH key and add it to the container's known keys
+        keyfile = 'id_{}'.format(self._container_name)
+        keyfile_path = os.path.join(self._tmp, keyfile)
+        if not os.path.exists(keyfile_path):
+            # Swallow output and allow exceptions to save stdout
+            check_output(['ssh-keygen', '-o', '-N', '', '-f', keyfile_path])
+
+        # Don't use os.path.join because a Windows client would use \
+        ssh_config = '/root/.ssh'
+        ssh_authorized_keys = ssh_config + '/authorized_keys'
+
+        self._container_run(['mkdir', '-p', ssh_config])
+        self._container_run(['chmod', '700', ssh_config])
+        self._container_run(['tee', '-a', ssh_authorized_keys],
+                            stdin=open('{}.pub'.format(keyfile_path), 'r'))
+        self._container_run(['chmod', '600', ssh_authorized_keys])
+
+        ssh_address = self._get_container_address()
+        logger.info('Connecting to {} via SSH'.format(ssh_address))
+        ssh_cmd = ['ssh', '-C', '-F', '/dev/null',
+                   '-o', 'IdentityFile={}'.format(keyfile_path),
+                   '-o', 'StrictHostKeyChecking=no',
+                   '-o', 'UserKnownHostsFile=/dev/null',
+                   '-o', 'User=root',
+                   '-p', '22', ssh_address]
+
+        # One-off command to verify the connection
+        try:
+            check_output(ssh_cmd + ['ls'])
+        except CalledProcessError as e:
+            raise SnapcraftEnvironmentError('Failed to setup SSH') from e
+
+        # Use sshfs in slave mode inside SSH to reverse mount destination
+        self._container_run(['apt-get', 'install', '-y', 'sshfs'])
+        self._container_run(['mkdir', '-p', destination])
+        # Pipes for sshfs and sftp-server to communicate
+        stdin1, stdout1 = os.pipe()
+        stdin2, stdout2 = os.pipe()
+        try:
+            self._host_run(['/usr/lib/sftp-server'],
+                           stdin=stdin1, stdout=stdout2)
+        except CalledProcessError as e:
```

Comment:
it seems I didn't get caught by this (same issue as that lxd bug you fixed, `FileNotFoundError`)
```
snapcraft
"grade" property not specified: defaulting to "stable"
Waiting for a network connection...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.5/urllib/request.py", line 163, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.5/urllib/request.py", line 466, in open
    response = self._open(req, data)
  File "/usr/lib/python3.5/urllib/request.py", line 484, in _open
    '_open', req)
  File "/usr/lib/python3.5/urllib/request.py", line 444, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.5/urllib/request.py", line 1282, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/usr/lib/python3.5/urllib/request.py", line 1257, in do_open
    r = h.getresponse()
  File "/usr/lib/python3.5/http/client.py", line 1197, in getresponse
    response.begin()
  File "/usr/lib/python3.5/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.5/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.5/socket.py", line 575, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out
Network connection established
Get:1 http://security.ubuntu.com/ubuntu xenial-security InRelease [102 kB]
Hit:2 http://archive.ubuntu.com/ubuntu xenial InRelease                   
Get:3 http://archive.ubuntu.com/ubuntu xenial-updates InRelease [102 kB]      
Get:4 http://archive.ubuntu.com/ubuntu xenial-backports InRelease [102 kB]    
Fetched 306 kB in 3s (85.0 kB/s)                                  
Reading package lists... Done
Mounting /home/sergiusens/source/snapcraft/demos/godd into container
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDquEdseWFnTTJkYdRHgpK8C6k+kSXVAgAll7dR5HAd0CWNg9Yfb0rqHCcj+YKMpPXd7hMTvMQ2BKdqimx52s055PEZab/GnFRALoXYLa/6i0P3CsYUA7FNqZAMBF5FVSjUIM+tqiwJNbGA7J7yAR/zmhOGMXYgKWxjcOnu8YGp72qOtS9tSZV5ZVovWjaAvOCBR519Khwnd123gH5L8XtXs2zP4ZDcVIgp0NXwSPOqWfDkykcFXgN3PpwI2ZtIdGTxSG/drTlnVjdTOnCINZ8gwSz5u+R+jerPQaT4dHb0iiJSV4cQ/gJYyLHsuX5nGODxIn3TwniqK0nhmjA1hp7x sergiusens@mirkwood
Connecting to 10.113.236.26 via SSH
Warning: Permanently added '10.113.236.26' (ECDSA) to the list of known hosts.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  gir1.2-gudev-1.0 golang-1.6-go golang-1.6-race-detector-runtime
  golang-1.6-src golang-go golang-race-detector-runtime golang-src
  libglib2.0-bin libglib2.0-dev libgudev-1.0-0 libgudev-1.0-dev libpcre16-3
  libpcre3-dev libpcre32-3 libpcrecpp0v5 pkg-config zlib1g-dev
Use 'apt autoremove' to remove them.
The following NEW packages will be installed:
  sshfs
0 upgraded, 1 newly installed, 0 to remove and 76 not upgraded.
Need to get 41.7 kB of archives.
After this operation, 138 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu xenial/universe amd64 sshfs amd64 2.5-1ubuntu1 [41.7 kB]
Fetched 41.7 kB in 2s (19.1 kB/s)
Selecting previously unselected package sshfs.
(Reading database ... 38315 files and directories currently installed.)
Preparing to unpack .../sshfs_2.5-1ubuntu1_amd64.deb ...
Unpacking sshfs (2.5-1ubuntu1) ...
Processing triggers for man-db (2.7.5-1) ...
Setting up sshfs (2.5-1ubuntu1) ...
Stopping shire:snapcraft-godd
Traceback (most recent call last):
  File "/home/sergiusens/venv/snapcraft/bin/snapcraft", line 11, in <module>
    load_entry_point('snapcraft', 'console_scripts', 'snapcraft')()
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/pkg_resources/__init__.py", line 561, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2631, in load_entry_point
    return ep.load()
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2291, in load
    return self.resolve()
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2297, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/sergiusens/source/snapcraft/snapcraft/cli/__main__.py", line 19, in <module>
    run(prog_name='snapcraft')
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 1043, in invoke
    return Command.invoke(self, ctx)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/sergiusens/source/snapcraft/snapcraft/cli/__init__.py", line 110, in run
    ctx.forward(lifecyclecli.commands['snap'])
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 553, in forward
    return self.invoke(cmd, **kwargs)
  File "/home/sergiusens/venv/snapcraft/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/sergiusens/source/snapcraft/snapcraft/cli/lifecycle.py", line 129, in snap
    lifecycle.containerbuild('snap', project_options, output, directory)
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lifecycle.py", line 354, in containerbuild
    metadata=config.get_metadata()).execute(step, args)
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lxd.py", line 409, in execute
    super().execute(step, args)
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lxd.py", line 144, in execute
    self._setup_project()
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lxd.py", line 317, in _setup_project
    self._ensure_mount(self._project_folder, self._source)
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lxd.py", line 332, in _ensure_mount
    return self._remote_mount(destination, source)
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lxd.py", line 382, in _remote_mount
    stdin=stdin1, stdout=stdout2)
  File "/home/sergiusens/source/snapcraft/snapcraft/internal/lxd.py", line 395, in _host_run
    self._processes += [Popen(cmd, **kwargs)]
  File "/usr/lib/python3.6/subprocess.py", line 707, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1333, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/lib/sftp-server'
```